### PR TITLE
Eva/tgsrv 264 line icons

### PR DESCRIPTION
**Pull request by gevos**

### CODE REVIEW

File: src/components/FeatureStyler/FeatureStyler.js [hunk 1]
```diff
@@ -497,6 +591,30 @@ class FeatureStyler extends PureComponent {
       textRotation = featStyle.getText().getRotation();
     }
 
+    const oldStyles = FeatureStyler.getStyleAsArray(feature);
+
+    if (oldStyles.length > 1) {
```
Comment (oterral) [inline src/components/FeatureStyler/FeatureStyler.js]: This 'if' seems useless

File: src/components/FeatureStyler/FeatureStyler.js [hunk 2]
```diff
@@ -497,6 +591,30 @@ class FeatureStyler extends PureComponent {
...
+      for (let i = 1; i < oldStyles.length; i += 1) {
+        if (oldStyles[i].getImage() && oldStyles[i].getImage().getSrc()) {
+          for (let j = 0; j < lineIcons.length; j += 1) {
```
Comment (oterral) [inline src/components/FeatureStyler/FeatureStyler.js]: Can you try to create a `findLineIcon(style, lineIcons)`   function ?

File: src/components/FeatureStyler/FeatureStyler.js [hunk 3]
```diff
@@ -497,6 +591,30 @@ class FeatureStyler extends PureComponent {
...
+              const coord = oldStyles[i].getGeometry().getCoordinates();
+              const startCoord = FeatureStyler.getVertexCoord(
+                feature.getGeometry(),
+              );
+              if (coord[0] === startCoord[0] && coord[1] === startCoord[1]) {
+                if (!lineStartIcon) {
+                  this.setState({ lineStartIcon: lineIcons[j].icon });
+                }
+              } else if (!lineEndIcon) {
+                this.setState({ lineEndIcon: lineIcons[j].icon });
```
Comment (oterral) [inline src/components/FeatureStyler/FeatureStyler.js]: `setState` is already called at the end, can you use this call with the lineStaertIcon lineEndIcon values

File: src/components/FeatureStyler/FeatureStyler.js (hunk 1 above)
Comment (gevos) [reply src/components/FeatureStyler/FeatureStyler.js]: If there aren't any extra styles defined, oldStyles has only a length of 1. Then I can't start iterating with 1.

File: src/components/FeatureStyler/__snapshots__/FeatureStyler.test.js.snap [hunk 4]
```diff
@@ -601,6 +601,96 @@ exports[`FeatureStyler matches snapshot when feature has a styleFunction() which
...
 >
+  <div
+    className="tm-modify-stroke"
+  >
+    <div>
+      Start:
+    </div>
+    <div
+      className="tm-button
+                "
```
Comment (oterral) [inline src/components/FeatureStyler/__snapshots__/FeatureStyler.test.js.snap]: Can we avoid this line creation ?

**Comment by oterral** [comment]
@gevos Do we nned this weird marker icon https://github.com/geops/react-spatial/pull/176/files?short_path=5e33f9b#diff-5e33f9bb148c79a17874b1bcd14c532d ? 

The marker icon doesn't seem to be well displayed on the example

File: src/components/FeatureStyler/FeatureStyler.js (hunk 1 above)
Comment (oterral) [reply src/components/FeatureStyler/FeatureStyler.js]: if oldStyles.length <= 1 , the code just doesn't go inside the loop    because i < arr.length

Comment (gevos) [reply src/components/FeatureStyler/FeatureStyler.js]: fixed

File: src/components/FeatureStyler/FeatureStyler.js (hunk 3 above)
Comment (gevos) [reply src/components/FeatureStyler/FeatureStyler.js]: done

**Comment by gevos** [comment]
> @gevos Do we nned this weird marker icon https://github.com/geops/react-spatial/pull/176/files?short_path=5e33f9b#diff-5e33f9bb148c79a17874b1bcd14c532d ?

fixed

> The marker icon doesn't seem to be well displayed on the example

As this is already wrong in the deployed version and due to lack of time, I didn't fix this.

File: src/components/FeatureStyler/FeatureStyler.js (hunk 2 above)
Comment (gevos) [reply src/components/FeatureStyler/FeatureStyler.js]: Done. However, now we have more line changes.

File: src/components/FeatureStyler/__snapshots__/FeatureStyler.test.js.snap (hunk 4 above)
Comment (gevos) [reply src/components/FeatureStyler/__snapshots__/FeatureStyler.test.js.snap]: Fixed

**Comment by mhaertwig** [comment]
**As this is already wrong in the deployed version and due to lack of time, I didn't fix this.**

@gevos, seriously?

**Comment by gevos** [comment]
> **As this is already wrong in the deployed version and due to lack of time, I didn't fix this.**
> 
> @gevos, seriously?

Yes, seriously, I have only a certain amount of time. ;-)

File: src/components/FeatureStyler/FeatureStyler.js [hunk 5]
```diff
@@ -497,6 +614,21 @@ class FeatureStyler extends PureComponent {
       textRotation = featStyle.getText().getRotation();
     }
 
+    const oldStyles = FeatureStyler.getStyleAsArray(feature);
+
+    const newLineStartIcon =
+      lineStartIcon ||
+      FeatureStyler.findLineIcon(oldStyles, feature.getGeometry(), lineIcons);
```
Comment (oterral) [inline src/components/FeatureStyler/FeatureStyler.js]: `findLineIcon` function updates the state of the component from the current style of the feature. This function doesn't care of the current state of the component. Imagine you click on another line the state of the component must be updated according to the new feature's style and ignoring what is the current state.

That's why you should remove the 'lineStartIcon || ' line.

Comment (gevos) [reply src/components/FeatureStyler/FeatureStyler.js]: Done.

File: src/components/FeatureStyler/FeatureStyler.js [hunk 6]
```diff
@@ -264,8 +291,32 @@ class FeatureStyler extends PureComponent {
...
+
+  static getLineIcon(geom, icon, color, start = true) {
+    const coordA = FeatureStyler.getVertexCoord(geom, start, 1);
+    const coordB = FeatureStyler.getVertexCoord(geom, start);
+    const dx = coordA[0] - coordB[0];
+    const dy = coordA[1] - coordB[1];
+    const rotation = Math.atan2(dy, dx);
+
+    return new Style({
+      geometry: new Point(coordB),
```
Comment (oterral) [inline src/components/FeatureStyler/FeatureStyler.js]: If you use a geometryFunction  , the arrow should move with the line properly

**Comment by oterral** [comment]
@gevos I've fixed the marker here https://github.com/geops/react-spatial/pull/178/files#diff-720c514dddba47247b577dc53a4134a2

File: src/components/FeatureStyler/FeatureStyler.js (hunk 6 above)
Comment (gevos) [reply src/components/FeatureStyler/FeatureStyler.js]: Good idea!
Done.

File: src/components/FeatureStyler/FeatureStyler.js [hunk 7]
```diff
@@ -348,7 +426,36 @@ class FeatureStyler extends PureComponent {
...
+            const coord = styles[i]
+              .getGeometry()(feature)
+              .getCoordinates();
+            const startCoord = FeatureStyler.getVertexCoord(geom);
+            const endCoord = FeatureStyler.getVertexCoord(geom, false);
+            if (
+              (coord[0] === startCoord[0] &&
+                coord[1] === startCoord[1] &&
+                start) ||
+              (coord[0] === endCoord[0] && coord[1] === endCoord[1] && !start)
```
Comment (oterral) [inline src/components/FeatureStyler/FeatureStyler.js]: I don't think we need to care about the coordinates, here, we just want to find which icon is used.

Comment (oterral) [reply src/components/FeatureStyler/FeatureStyler.js]: of course we did, forget it

**Comment by oterral** [comment]
@gevos i'merge master and cleaning some props to use the same properties as for others icons.
I also add the read/write for KML export/import, it should work.

**Comment by gevos** [comment]
When the arrow points to the line, the line intersects the arrow a little bit.

**Pull request merged**